components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/prompt/GeckoPromptDelegate.kt

```diff
@@ -0,0 +1,178 @@
+import org.mozilla.geckoview.AllowOrDeny
+import org.mozilla.geckoview.GeckoResult
+import org.mozilla.geckoview.GeckoSession
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.AlertCallback
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.AuthCallback
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.AuthOptions
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.ButtonCallback
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.Choice.CHOICE_TYPE_MENU
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.Choice.CHOICE_TYPE_MULTIPLE
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.Choice.CHOICE_TYPE_SINGLE
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.ChoiceCallback
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.FileCallback
+import org.mozilla.geckoview.GeckoSession.PromptDelegate.TextCallback
+
+typealias GeckoChoice = GeckoSession.PromptDelegate.Choice
+
+/**
+ * Gecko-based PromptDelegate implementation.
+ */
+@Suppress("TooManyFunctions")
+class GeckoPromptDelegate(private val geckoEngineSession: GeckoEngineSession) :
+    GeckoSession.PromptDelegate {
+
+    override fun onChoicePrompt(
+        session: GeckoSession?,
+        title: String?,
+        msg: String?,
+        type: Int,
+        geckoChoices: Array<out GeckoChoice>,
+        callback: ChoiceCallback
+    ) {
+
+        val pair = convertToChoices(geckoChoices)
+        // An array of all the GeckoChoices transformed as local Choices object.
+        val choices = pair.first
+        // A map that contains all local choices and map to GeckoChoices
+        val mapChoicesToGeckoChoices = pair.second
+
+        when (type) {
+
+            CHOICE_TYPE_SINGLE -> {
+                geckoEngineSession.notifyObservers {
+                    onPromptRequest(SingleChoice(choices) { selectedChoice ->
+                        val geckoChoice = mapChoicesToGeckoChoices[selectedChoice]
+                        callback.confirm(geckoChoice)
+                    })
+                }
+            }
+
+            CHOICE_TYPE_MENU -> {
+                geckoEngineSession.notifyObservers {
+                    onPromptRequest(MenuChoice(choices) { selectedChoice ->
+                        val geckoChoice = mapChoicesToGeckoChoices[selectedChoice]
+                        callback.confirm(geckoChoice)
+                    })
+                }
+            }
+
+            CHOICE_TYPE_MULTIPLE -> {
+                geckoEngineSession.notifyObservers {
+                    onPromptRequest(MultipleChoice(choices) { choices ->
+                        val ids = choices.toIdsArray()
+                        callback.confirm(ids)
+                    })
+                }
+            }
+        }
+    }
+
+    override fun onButtonPrompt(
+        session: GeckoSession?,
+        title: String?,
+        msg: String?,
+        btnMsg: Array<out String>?,
+        callback: ButtonCallback?
+    ) {
+    }
```

Comment:
nit: `= Unit` avoids you to have empty blocks :)

Reply:
Yeah, this will be implemented in coming PRs. How should I handle it ?

Reply:
Yeah, that's fine. `) = Unit` looks more like it is an empty implementation on purpose instead of "Oops, forgot the code" :D But feel free to land as-is or throw in a comment linking to the issue if there's one already.

Reply:
👍